CONTRIBUTING.md

```diff
@@ -0,0 +1,127 @@
+
+You should hook a function whe you want to alter the behavior of the function. If you don't need to alter the behavior, use FIND_PATTERN.
+
+## FIND_PATTERN - macro
+Example:
+```
+// hud.cpp
+FIND_PATTERN(vguimatsurface, StartDrawing);
+```
+```
+// patterns.hpp
+...
+    PATTERNS(
+        StartDrawing,
+        "5135",
+        "55 8B EC 83 E4 C0 83 EC 38 80 ?? ?? ?? ?? ?? ?? 56 57 8B F9 75 57 8B ?? ?? ?? ?? ?? C6 ?? ?? ?? ?? ?? ?? FF ?? 8B 10 6A 00 8B C8 8B 42 20");
+...
+```
+FIND_PATTERN finds a pointer to the address where the pattern is found. This pointer is placed into ORIG_StartDrawing in this case. This is typically used for finding a function and then the pointer can be used to call the function.
+
+## What pattern found my function?
+In some cases you have multiple patterns under the same name to support multiple versions of the game. You might also want to do some pattern specific setup, to account for differences between versions. To determine the pattern used, you can use the function GetPatternIndex. Example:
+```
+// patterns.hpp:
+PATTERNS(
+    SpawnPlayer,
+    "5135",
+    "83 EC 14 80 3D ?? ?? ?? ?? 00 56 8B F1 74 ?? 6A 00 B9 ?? ?? ?? ?? E8 ?? ?? ?? ?? EB 23 8B 0D ?? ?? ?? ?? 8B 01 8B 96 ?? ?? ?? ?? 8B 40 0C 52 FF D0 8B 8E ?? ?? ?? ?? 51 E8 ?? ?? ?? ?? 83 C4 04 8B 46 0C 8B 56 04 8B 52 74 83 C0 01",
+    "5339",
+    "55 8B EC 83 EC 14 80 3D ?? ?? ?? ?? 00 56 8B F1 74 ?? 6A 00 B9 ?? ?? ?? ?? E8 ?? ?? ?? ?? EB 23 8B 0D ?? ?? ?? ?? 8B 01 8B 96 ?? ?? ?? ?? 8B 40 0C 52 FF D0 8B 8E ?? ?? ?? ?? 51 E8 ?? ?? ?? ?? 83 C4 04 8B 46 0C 8B 56 04 8B 52 74 40",
+...
+```
+```
+// pause.cpp
+	if (ORIG_SpawnPlayer)
+	{
+		int ptnNumber = GetPatternIndex((void**)&ORIG_SpawnPlayer);
+
+		switch (ptnNumber)
+		{
+		case 0:
+			pM_bLoadgame = (*(bool**)(ORIG_SpawnPlayer + 5));
+			pGameServer = (*(void**)(ORIG_SpawnPlayer + 18));
+			break;
+...
+```
+Here the 0 index corresponds to the 5135 pattern, 1 to 5339 and so on.
+
+# Initializing console commands/variables
+Console commands and variables have to be initialized for SPT to load them. Otherwise they get automatically unloaded. Example:
+```
+// ipc-spt.cpp
+void ipc::IPCFeature::LoadFeature()
+{
+	if (FrameSignal.Works)
+	{
+		Init();
+		FrameSignal.Connect(Loop);
+#ifndef OE
+		InitCommand(y_spt_ipc_ent);
+		InitCommand(y_spt_ipc_properties);
+#endif
+		InitCommand(y_spt_ipc_echo);
+		InitCommand(y_spt_ipc_playback);
+		InitCommand(y_spt_ipc_gamedir);
+
+		InitConcommandBase(y_spt_ipc);
+		InitConcommandBase(y_spt_ipc_port);
+	}
+}
+```
+The FrameSignal.Works check checks if the FrameSignal has been initialized properly. If it is, then all the relevant commands are initialized with the macro InitCommand and variables are initialized with the command InitConcommandBase.
```

Comment:
I'd also use inline code where appropriate throughout the document, e.g. here `FrameSignal.Works`